src/dbnode/storage/namespace/schema.go

```diff
@@ -0,0 +1,212 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package namespace
+
+import (
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"io/ioutil"
+	"sort"
+
+	"github.com/golang/protobuf/proto"
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+	nsproto "github.com/m3db/m3/src/dbnode/generated/proto/namespace"
+	xerrors "github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/ident"
+)
+
+var (
+	errInvalidSchema  = errors.New("invalid schema definition")
+	errSchemaNotFound = errors.New("schema is not found")
+)
+
+type schemaDescr struct {
+	id      ident.ID
+	version uint32
+	md      *desc.MessageDescriptor
+}
+
+func (s *schemaDescr) ID() ident.ID {
+	return s.id
+}
+
+func (s *schemaDescr) Version() uint32 {
+	return s.version
+}
+
+func (s *schemaDescr) Equal(o SchemaDescr) bool {
+	if s == nil && o == nil {
+		return true
+	}
+	if s != nil && o == nil || s == nil && o != nil {
+		return false
+	}
+	return s.ID().String() == o.ID().String() && s.Version() == o.Version()
+}
+
+func (s *schemaDescr) Get() *desc.MessageDescriptor {
+	return s.md
+}
+
+func (s *schemaDescr) String() string {
+	if s.md == nil {
+		return ""
+	}
+	return s.md.String()
+}
+
+type schemaRegistry struct {
+	options *nsproto.SchemaOptions
+	schemas map[string]*schemaDescr
+}
+
+func (sr *schemaRegistry) Equal(o SchemaRegistry) bool {
+	var osr *schemaRegistry
+	var ok bool
+	if sr == nil && o == nil {
+		return true
+	}
+	if sr != nil && o == nil || sr == nil && o != nil {
+		return false
+	}
+
+	if osr, ok = o.(*schemaRegistry); !ok {
+		return false
+	}
+	if len(sr.schemas) != len(osr.schemas) {
+		return false
+	}
+	for n, sd := range sr.schemas {
+		var osd *schemaDescr
+		if osd, ok = osr.schemas[n]; !ok {
+			return false
+		}
+		if !sd.Equal(osd) {
+			return false
+		}
+	}
+	return true
+}
+
+func (sr *schemaRegistry) Get(id ident.ID) (SchemaDescr, error) {
+	sd, ok := sr.schemas[id.String()]
+	if !ok {
+		return nil, errSchemaNotFound
+	}
+	return sd, nil
+}
+
+func (sr *schemaRegistry) IDs() []ident.ID {
+	ids := make([]ident.ID, len(sr.schemas))
+	i := 0
+	for id := range sr.schemas {
+		ids[i] = ident.StringID(id)
+		i++
+	}
+	return ids
+}
+
+// toSchemaOptions returns the corresponding SchemaOptions proto for the provided SchemaRegistry
+func toSchemaOptions(sr SchemaRegistry) *nsproto.SchemaOptions {
+	if sr == nil {
+		return nil
+	}
+	_, ok := sr.(*schemaRegistry)
+	if !ok {
+		return nil
+	}
+	return sr.(*schemaRegistry).options
+}
+
+// LoadSchemaRegistry loads schema registry from SchemaOptions proto.
+func LoadSchemaRegistry(options *nsproto.SchemaOptions) (SchemaRegistry, error) {
+	if options.GetRepo() == nil || len(options.GetRepo().GetHistory()) == 0 ||
+		len(options.GetSchemas()) == 0 {
+		return nil, nil
```

Comment:
Would it be less confusing if we returned an "empty" schema registry instead?

Reply:
agree, changed default from nil to empty